client/hotrod-client/src/test/java/org/infinispan/client/hotrod/tx/RecoveryTest.java

```diff
@@ -0,0 +1,247 @@
+      for (EmbeddedCacheManager cm : cacheManagers) {
+         //use the same time service in all managers
+         replaceComponent(cm, TimeService.class, timeService, true);
+         //stop reaper. we are going to trigger it manually
+         extractGlobalComponent(cm, GlobalTxTable.class).stop();
+      }
+      defineInAll(cacheName(), cacheBuilder);
+   }
+
+   @Override
+   protected org.infinispan.client.hotrod.configuration.ConfigurationBuilder createHotRodClientConfigurationBuilder(
+         int serverPort) {
+      org.infinispan.client.hotrod.configuration.ConfigurationBuilder clientBuilder = super
+            .createHotRodClientConfigurationBuilder(serverPort);
+      clientBuilder.forceReturnValues(false);
+      clientBuilder.transaction().transactionManagerLookup(RemoteTransactionManagerLookup.getInstance());
+      clientBuilder.transaction().transactionMode(TransactionMode.FULL_XA);
+      clientBuilder.transaction().timeout(10, TimeUnit.SECONDS);
+      return clientBuilder;
+   }
+
+   private void assertRecoveryIteration(XAResource xaResource, Xid local, Xid remote) throws Exception {
+      Xid[] actual = xaResource.recover(XAResource.TMSTARTRSCAN);
+      assertTrue(actual.length != 0);
+      if (actual.length == 1) {
+         //it returned only the local transaction
+         assertEquals(local, actual[0]);
+         actual = xaResource.recover(XAResource.TMENDRSCAN);
+         assertEquals(1, actual.length); //other client transaction
+         assertEquals(remote, actual[0]);
+      } else {
+         //the server replied quick enough
+         assertEquals(local, actual[0]);
+         assertEquals(remote, actual[1]);
+         actual = xaResource.recover(XAResource.TMENDRSCAN);
+         assertEquals(0, actual.length);
+      }
+   }
+
+   private void assertBeforeTimeoutRecoveryIteration(XAResource xaResource, Xid local) throws Exception {
+      Xid[] actual = xaResource.recover(XAResource.TMSTARTRSCAN);
+      assertEquals(1, actual.length);
+      assertEquals(local, actual[0]);
+      actual = xaResource.recover(XAResource.TMENDRSCAN);
+      assertEquals(0, actual.length);
+   }
+
+   private void prepare(int index) throws Exception {
+      RemoteTransactionManager tm = remoteTM(index);
+      TransactionImpl tx = (TransactionImpl) tm.getTransaction();
+      tm.suspend();
+      assertTrue(tx.runPrepare());
+   }
+
+   private Xid xid(int index) {
+      TransactionImpl tx = (TransactionImpl) remoteTM(index).getTransaction();
+      return tx.getXid();
+   }
+
+   private XAResource xaResource(int index) throws Exception {
+      RemoteTransactionManager tm = remoteTM(index);
+      tm.begin();
+      RemoteCache<String, String> cache = remoteCache(index);
+      cache.put("_k_", "_v_");
+      TransactionImpl tx = (TransactionImpl) tm.getTransaction();
+      XAResource xaResource = tx.getEnlistedResources().iterator().next();
+      tm.commit();
+      xaResource.forget(tx.getXid());
```

Comment:
I really think `recover()` should work without a dummy transaction, e.g. immediately after a client restart.

Reply:
you still need an XaResource and there is no XaResource factories.